Named external requires - fixes #5

bclinkinbeard opened the pull request:
This allows commonly used dependencies to be exposed as a named require, eliminating the need for repetitive, error prone and potentially lengthy relative pathing. Example syntax in the Grunt file looks like this.

```
exposed: {
    Clazz: './libs/Class.js'
}
```

which in turn enables requires like this

```
require('Clazz');
```

bclinkinbeard committed:
Add support for named external requires

This allows commonly used dependencies to be exposed as a named require, eliminating the need for repetitive, error prone and potentially lengthy relative pathing. Example syntax in the Grunt file looks like this.

```
exposed: {
	Clazz: './libs/Class.js'
}
```
which in turn enables requires like this
```
require('Clazz');
```

shanejonas commented on tasks/browserify2.coffee:
```diff
@@ -11,6 +11,7 @@ module.exports = (grunt)->
     targetConfig = config[@target]
     {entry, mount, server, debug, compile, beforeHook} = targetConfig
     bundle = browserify entry
+    bundle.require( val, { expose: key } ) for key, val of exposed
```
where is `exposed` coming from here? You need to pull it out of the `targetConfig` above.

``` coffee
{entry, mount, server, debug, compile, beforeHook, exposed} = targetConfig
```

bclinkinbeard replied:
Whoops, I had that change locally but forgot to make it when creating the
pull request. (I just edited the file on the site because I haven't
actually cloned my fork yet.)

It also occurred to me that having these above the target level would be
better since you could then avoid having to duplicate them for every
target. Will look into how to do that tonight and submit a corrected PR.

bclinkinbeard committed:
Add support for both task and target level expose options

bclinkinbeard commented:
This should now be a proper implementation supporting both task and target level options objects as described in the Grunt docs. This would most likely only be used at the task level, but adding target level support was trivial so why not. Also note the option name was changed to `expose`, to match the option named used internally by browserify.

Example syntax:

```
browserify2: {
    options: {
        expose: {
            Clazz: './libs/Class.js' // enables require('Clazz') for all targets
        }
    },
    dev: {
        options: {
            expose: {
                Cls: './libs/Class.js' // enables require('Cls') but only for dev target. not recommended.
            }
        },
        entry: './src/mtlg.js',
        compile: './dist/application.js',
        debug: true
    },
    compile: {
        entry: './src/mtlg.js',
        compile: './dist/application.js'
    }
}
```

Using the syntax above, both require('Cls') and require('Clazz') would be valid when running the dev target.

bclinkinbeard commented:
Now having second thoughts about supporting this at the target level as I feel it just opens the door to unintended behavior. Unless you feel strongly otherwise, refrain from merging and I will update later to limit this to task level.

OliverJAsh commented:
I think the target level has use cases. I.e. changing a module depending on environment.

bclinkinbeard commented:
Good point. I was only thinking about changing the key but changing
the value could definitely be useful.

bclinkinbeard commented:
Looks like using expose is OK, at least according to browserify-shim: https://github.com/thlorenz/browserify-shim/tree/master/examples/shim-underscore

shanejonas commented:
So you can use this is instead of `browserify-shim`?

bclinkinbeard commented:
No, it's not really a replacement for shimming (whose main purpose, as I understand it, is to make non-CommonJS modules Browserify friendly), it simply lets you specify a custom name/path with which you can refer to a module, rather than having to use the actual relative file path. For example, I think it would be handy to be able to just do `require("libs/_")` from anywhere, instead of something like `require("./../../../../libs/underscore")` from some deeply nested module.

Thoughts?

bclinkinbeard committed:
Added support for file globbing in expose property

If expose contains a files array, member objects will be passed to
grunt.file.expandMapping() to support exposing patterns of files
as simplified names/paths. Each object in the files array must
provide the cwd and src properties, while dest is optional.

For example, the following config will expose all .js files for ```require```ing within the vendor directory, at any level, simply by their filename, minus the extension.

```
expose: {
	files: [
		{
			cwd: './vendor/',
			src: ['**/*.js']
		}
	]
}
```

Adding ```dest: 'lib/'``` to the object would enable ```require("lib/somename")```

bclinkinbeard commented:
OK, so I can't seem to get it through my head that markdown in commit messages doesn't get rendered, so I am repeating the comment here. Basically, this latest commit makes it so you could easily expose a whole set of files, rather than having to specify each one individually. The use case that drove the feature was to allow me to expose all of the 3rd party libraries I am using by name, so `require('angular')` for `libs/angular/angular.js`, `require('underscore')` for `libs/underscore.js`, etc. Commit message follows.

If expose contains a files array, member objects will be passed to
grunt.file.expandMapping() to support exposing patterns of files
as simplified names/paths. Each object in the files array must
provide the cwd and src properties, while dest is optional.

For example, the following config will expose all .js files for `require`ing within the vendor directory, at any level, simply by their filename, minus the extension.

```
expose: {
    files: [
        {
            cwd: './vendor/',
            src: ['**/*.js']
        }
    ]
}
```

Adding `dest: 'lib/'` to the object would enable `require("lib/somename")`

bclinkinbeard commented:
Confirmation expose is there to be used: https://github.com/substack/node-browserify/pull/350

OliverJAsh commented:
@shanejonas Can we get this merged?

shanejonas commented:
this and  #11 are now in v0.1.5

shanejonas commented:
@bclinkinbeard could you make a pull request with a usage example for the README?
